src/main/scala/com/github/codelionx/dodo/actors/ODMaster.scala

```diff
@@ -296,116 +293,189 @@ class ODMaster(inputFile: Option[File])
       val sum = sortedWorkloads.map(_._1).sum
       val averageWl: Int = sum / (sortedWorkloads.size + 1)
+      val pendingSum = sortedWorkloads.map(_._2).sum
       var ownWorkLoad = 0
 
       log.debug(
-        "Work stealing status: averageWl={}, our hasPendingWork ODs={}",
+        "Work stealing status: averageWl={}, others' pendingSum={}, our pendingWork ODs={}",
         averageWl,
+        pendingSum,
         candidateQueue.pendingSize
       )
-      if (sum > 0) {
+      if (averageWl > 0) {
         // there is work to steal, steal some from the busiest nodes
-        for ((otherSize, otherRef) <- sortedWorkloads) {
-          val amountToSteal = Seq(
-            otherSize - averageWl,
-            averageWl - ownWorkLoad,
-            settings.workers * settings.maxBatchSize
-          ).min
-          if (amountToSteal > 0) {
-            otherRef ! WorkToSend(amountToSteal)
-            log.info("Stealing {} elements from {}", amountToSteal, otherRef)
-            ownWorkLoad += amountToSteal
-          }
+        val newPendingResponses = sortedWorkloads.flatMap {
+          case (otherSize, _, otherRef) =>
+            val amountToSteal = Seq(
+              otherSize - averageWl,
+              averageWl - ownWorkLoad,
+              settings.workers * settings.maxBatchSize
+            ).min
+            if (amountToSteal > 0) {
+              otherRef ! WorkToSend(amountToSteal)
+              log.info("Stealing {} elements from {}", amountToSteal, otherRef)
+              ownWorkLoad += amountToSteal
+              context.watch(otherRef)
+              Some(otherRef)
+            } else {
+              None
+            }
         }
+        context.become(workStealing(table, newPendingResponses.toSet))
       } else if (candidateQueue.hasNoPendingWork) {
-        // we have no work left and got no work from our work stealing attempt, finished?
-        // TODO: make sure others don't have anything pending work anymore as well
-        log.warning(
-          "This node has no more pending candidates and the queue of other nodes is empty as well. Shutting down node."
-        )
-        shutdown()
+        // we have no work left and got no work from our work stealing attempt
+        if (pendingSum > 0) {
+          // others will soon have work for us to steal again
+          log.info("No work received, will ask again in three seconds")
+          context.system.scheduler.scheduleOnce(3 second, self, startWorkStealing(table))
+        }
+        // check if ALL nodes are completely out of work
+        startDowningProtocol(table)
       } else {
         // nothing to steal, but we still wait for worker results, wait for them
         log.info("Nothing to steal, but our workers are still busy.")
         context.become(findingODs(table))
       }
 
+    case Terminated(otherMaster) =>
+      val newPendingResponses = pendingResponses - otherMaster
+      context.become(workStealing(table, newPendingResponses))
```

Comment:
`context.become()` takes only effect after the `receive`-func was left. As you have other calls to `context.become()` below this and even call it in the `startDowningProtocol()` function, this call is a no-op in those two branches.

Can you move it into the top-level `else`-branch for easier comprehension?

Reply:
I've just noticed that you do it lake that in the `StolenWork()` case handler already.